src/twisted/internet/test/test_base.py

```diff
@@ -155,6 +158,45 @@ def fakeGetHostByName(name):
         result.put(IOError("The I/O was errorful"))
 
 
+    def test_resolverGivenStr(self):
+        """
+        L{ThreadedResolver.getHostByName} is passed L{str}, encoded using IDNA
+        if required.
+        """
+        calls = []
+
+        @implementer(IResolverSimple)
+        class FakeResolver(object):
+            def getHostByName(self, name, timeouts=()):
+                calls.append(name)
+                return Deferred()
+
+        class JustEnoughReactor(ReactorBase):
+            def installWaker(self):
+                pass
+
+        fake = FakeResolver()
+        reactor = JustEnoughReactor()
+        reactor.installResolver(fake)
+        rec = FirstOneWins(Deferred())
+        reactor.nameResolver.resolveHostName(
+            rec, u"example.example")
+        reactor.nameResolver.resolveHostName(
+            rec, "example.example")
+        reactor.nameResolver.resolveHostName(
+            rec, u"v\xe4\xe4ntynyt.example")
+        reactor.nameResolver.resolveHostName(
+            rec, u"\u0440\u0444.example")
```

Comment:
Should these again be `.invalid`?  Also it might be a good idea include an IDNA encoded native string, e.g. `" xn----7sbb4ac0ad0be6cf.xn--p1ai"`, so that this also serves as a regression test.